From 34a78e878a5641a4c6a1f5eda3735ef28dd7954b Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Fri, 13 Aug 2010 03:43:22 +0100
Subject: [PATCH 1/2] fs/buffer.c: Avoid ABI change in 2.6.32.19

The operation type flags used by ll_rw_block() and submit_bh() are
supposed to be interchangeable with those used by submit_bio(), but
RWA_MASK was not equal to 1 << BIO_RW_AHEAD.  Changing its value is
an ABI change; we need to accept the old value too.
---
 fs/buffer.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/fs/buffer.c b/fs/buffer.c
index 6fa5302..f24989c 100644
--- a/fs/buffer.c
+++ b/fs/buffer.c
@@ -2954,6 +2954,10 @@ int submit_bh(int rw, struct buffer_head * bh)
 	BUG_ON(buffer_delay(bh));
 	BUG_ON(buffer_unwritten(bh));
 
+	/* Fix broken value of RWA_MASK */
+	if (rw & 2)
+		rw ^= 2 | RWA_MASK;
+
 	/*
 	 * Mask in barrier bit for a write (could be either a WRITE or a
 	 * WRITE_SYNC
@@ -3027,6 +3031,10 @@ void ll_rw_block(int rw, int nr, struct buffer_head *bhs[])
 {
 	int i;
 
+	/* Fix broken value of RWA_MASK */
+	if (rw & 2)
+		rw ^= 2 | RWA_MASK;
+
 	for (i = 0; i < nr; i++) {
 		struct buffer_head *bh = bhs[i];
 
-- 
1.7.1

